syntax = "proto3";

import "google/protobuf/timestamp.proto";

import "model.proto";
import "history.proto";

package pd.transit;

message AuthRequest {
    string username = 1;
    string password = 2;
}

message AuthResponse {
    string token = 1;
}


message ReadUserRequest {
    string token = 1;
    string user_id = 2;
}

message ReadUserResponse {
    pd.model.User user = 1;
}

message CreateUserRequest {
    string token = 1;
}

message CreateUserResponse {
    pd.model.User user = 1;
}

message UpdateUserRequest {
    string token = 1;
    pd.model.User user = 2;
}

message UpdateUserResponse {
}

message CreateRegularMomentRequest {
    string token = 1;
    string user_id = 2;
    uint32 series = 3;
}

message CreateRegularMomentResponse {
    pd.history.RegularMoment moment = 1;
}

message ReadRegularMomentRequest {
    string token = 1;
    string user_id = 2;
    uint32 series = 3;
    google.protobuf.Timestamp timestamp = 4;
}

message ReadRegularMomentResponse {
    pd.history.RegularMoment moment = 1;
}

message UpdateRegularMomentRequest {
    string token = 1;
    string user_id = 2;
    pd.history.RegularMoment moment = 3;
}

message UpdateRegularMomentResponse {
}

service StorageService {
    rpc get_token(AuthRequest) returns (AuthResponse);

    // User methods
    rpc create_user(CreateUserRequest) returns (CreateUserResponse);
    rpc read_user(ReadUserRequest) returns (ReadUserResponse);
    rpc update_user(UpdateUserRequest) returns (UpdateUserResponse);

    // Regular methods
    rpc create_regular_moment(CreateRegularMomentRequest) returns (CreateRegularMomentResponse);
    rpc read_regular_moment(ReadRegularMomentRequest) returns (ReadRegularMomentResponse);
    rpc update_regular_moment(UpdateRegularMomentRequest) returns (UpdateRegularMomentResponse);
}
